docs/rfcs/20220804-ptensor/README.md

```diff
@@ -0,0 +1,260 @@
+
+## Proposal
+We propose two new Dialects and one dialect extension:
+1. __PTensor__ dialect: providing a tensor type (ptensor) and operations on tensors
+2. __Dist__ dialect: dealing with the aspects of distributed data management, such as distributed memory allocation, GC and communication
+3. __intel_sycl.device_region__ operation in __intel_sycl__ dialect (RFC #291)
+
+Additionally we propose appropriate passes
+1. Converting __PTensor__ operations into dialects __Linalg__, __GPU__, __intel_sycl__ and __Dist__
+2. Converting __Dist__ operations into appropriate runtime calls
+3. Converting __intel-sycl.device_region__ to appropriate runtime calls.
+
+### ptensor Type
+Logically, the ptensor type extends the `mlir::tensor` type with `device`, `team` and `gid` attributes. The `ptensor.ptensor` enables SSA values representing tensors with annotations for distributed and device operation. The tensors themselves are assumed to eventually lower to memrefs.
+
+The optional `device` indicates where the tensor lives, e.g. on which device. The target device is represented as a plain string which will be forwarded to the gpu/distributed runtimes. The exact syntax of device-identifiers is defined by the GPU runtime. For example, SYCL runtime will accept SYCL filter strings as defined [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md#sycl_device_filter). The default `NoneType` which disables device support.
+
+The optional `team` attribute indicates a team (of processes) among which the tensor is partitioned-distributed. The type of the `team` attribute depends on the underlying runtime; for MPI and Intel's distributed runtime this would be a `int64`. It defaults to `NoneType` which disables support for distributed operation.
+
+Notice: By default device and distribution support is disabled and so renders conventional host operations.
+
+### __PTensor__ Operations
+The initial set of operations matches the requirements of the core of [array-API](https://data-apis.org/array-api/latest/API_specification/index.html). Notice, since we focus on compute-follows-data, only the creation functions/operations will require the `device` and `team` attributes. The operations in the PTensor dialect accept different tensor-like types as input arguments: ptensors, tensors and MemRefs. Apparently, operations can only implement compute-follows-data if the inputs have `device` and `team` attributes, e.g. if they are of ptensor type. Standard tensor and memref inputs are treated as ptensors with default `device` and `team`.
```

Comment:
Actually, I don't like the idea of ptensor ops accepting usual tensors/memref arguments, better to have separate `to/from_tensor` and `to/from_memref` ops, which can also be useful outside this context.

Reply:
Good idea.

I understand fromTensor and from fromMemref. The other way round raises some questions. If the ptensor is a device tensor and/or distributed, how would suggest that toMemref/toTensor behave? Raise an error?

Maybe we do not need tomemref/tensor at all.

Reply:
I don't see any problem in this, we want to lower ptensor to usual tensor/memref eventually anyway, and at this point those ops will be just noop. We can left this question up to the lowering passes, so they can choose, if they want to lower them or just return error.

Reply:
Yes, we do not need to provide an implementation. Still, we need to define the semantics of the operation. The input to these operation is a ptensor, and the general expectation in this RFC is compute-follows-data. ToMemref and toTensor  effectively remove the location information. So what does this mean? The result lives in the default location (and so potentially needs to get copied over)? At least to me this does not make sense. And I do not see any other reasonable behavior.

If we can't define the semantics we better not have tomemref or totensor.

Reply:
Ok, I don't need them now for my cases. We can decide and add them later if needed.

Reply:
OK.

I added a `from_ranked` operation and changed the description accordingly.